sdk/eidolon_ai_sdk/agent/simple_agent.py

```diff
@@ -73,15 +84,83 @@ def make_output_schema(self, agent, handler):
         return str if self.output_schema == "str" else schema_to_model(self.output_schema, model_name)
 
 
-class SimpleAgentSpec(AgentSpec):
+class NamedCPU(BaseModel):
+    title: Optional[str] = None
+    cpu: AnnotatedReference[AgentCPU]
+    default: bool = False
+
+
+class SimpleAgentSpec(BaseModel):
     description: Optional[str] = None
     system_prompt: str = "You are a helpful assistant"
+    agent_refs: List[str] = []
     actions: List[ActionDefinition] = [ActionDefinition()]
+    cpu: AnnotatedReference[AgentCPU] = None
+    cpus: Optional[List[NamedCPU]] = []
+    title_generation_mode: Literal["none", "on_request"] = "on_request"
+
+    @model_validator(mode="before")
+    def validate_cpu(cls, value):
+        if "cpu" in value and "cpus" in value:
+            raise ValueError("Cannot specify both cpu and cpus")
+        return value
+
+    # noinspection PyTypeChecker
+    @model_validator(mode="after")
+    def validate_cpus(self):
+        if self.cpus:
+            self.cpu = None
+        return self
 
 
-class SimpleAgent(Agent, Specable[SimpleAgentSpec]):
+class SimpleAgent(Specable[SimpleAgentSpec]):
+    generate_title_prompt = ("You are generating a title for a conversation. Consider the context and content of the discussion or text. "
+                             "Create a concise, relevant, and accurate representation of the main topic or theme in the content."
+                             "Create a title that draws insiration from key phrases or ideas in the content. "
+                             "The title should be no longer than 5 words. Do not wrap the title in quotes. Answer only with the title. The prompt for the conversation is:\n")
+
     def __init__(self, spec):
         super().__init__(spec=spec)
+        if self.spec.cpu:
+            self.cpu = self.spec.cpu.instantiate()
+            self.cpu.title = self.cpu.__class__.__name__
+            self._register_refs_logic_unit(self.cpu, self.spec.agent_refs)
+        else:
+            self.cpus = []
+            for cpu_spec in self.spec.cpus:
+                cpu = cpu_spec.cpu.instantiate()
+                # todo - add title from metadata
+                cpu.title = cpu_spec.title or cpu.__class__.__name__
+                if cpu_spec.default:
+                    cpu.default = True
+                    self.cpu = cpu
+                self._register_refs_logic_unit(cpu, self.spec.agent_refs)
+                self.cpus.append(cpu)
+
+        if self.spec.title_generation_mode == "on_request":
+            # add a title generation action
+            action = ActionDefinition(
+                name="generate_title",
+                description="Generate a title for the conversation",
+                user_prompt=self.generate_title_prompt + "{{ body }}",
```

Comment:
Nit: why add body? Shouldn't we just let this come through generate_title_prompt if desired?

Reply:
The body is the question.  I didn't want to expose the prompt to the UI that generates the title.